checker/src/org/checkerframework/checker/genericeffects/GenericEffectVisitor.java

```diff
@@ -148,69 +138,958 @@ public Void visitMethod(MethodTree node, Void p) {
+     * @param warningMsg The warning message to be reported.
+     */
+    private void checkEnclosingMethod(Tree node, Class<? extends Annotation> targetEffect, String failureMsg, String warningMsg) {
+        MethodTree callerTree = TreeUtils.enclosingMethod(getCurrentPath());
+        ExecutableElement callerElt = TreeUtils.elementFromDeclaration(callerTree);
+        Class<? extends Annotation> callerEffect = atypeFactory.getDeclaredEffect(callerElt);
+        if (checker.getOption("ignoreEffects") != null)
+            targetEffect = extension.checkIgnoredEffects(checker.getOption("ignoreEffects"), targetEffect);
+        if(warningMsg != null) {
+            if(checker.getOption("ignoreWarnings") == null)
+                checker.report(Result.warning(warningMsg, targetEffect, callerEffect), node);
+            else if (checker.getOption("ignoreWarnings") != null && !extension.isIgnored(checker.getOption("ignoreWarnings"), warningMsg))
+                checker.report(Result.warning(warningMsg, targetEffect, callerEffect), node);
+        }
+        if (!genericEffect.LE(targetEffect, callerEffect)) {
+            if(checker.getOption("ignoreErrors") == null)
+                checker.report(Result.failure(failureMsg, targetEffect, callerEffect), node);
+            else if (checker.getOption("ignoreErrors") != null && !extension.isIgnored(checker.getOption("ignoreErrors"), failureMsg))
+                checker.report(Result.failure(failureMsg, targetEffect, callerEffect), node);
+        }
+    }
+
+    /**
+     * Method that can be used in a visitor method to see if a node is enclosed by a variable.
+     *
+     * @return A boolean representing whether the node is enclosed by a variable (true) or not (false).
+     */
+    private boolean hasEnclosingVariable() {
+        VariableTree varTree = TreeUtils.enclosingVariable(getCurrentPath());
+        return varTree != null;
+    }
+
+    /**
+     * Method that is used by all visitors to enforce the checker's lattice and report errors and warnings depending on the
+     * developers specifications. Compiler arguments are also checked to determine what should be ignored. This method
+     * checks at the variable level.
+     *
+     * @param node A tree node encountered during checking.
+     * @param targetEffect The effect of the tree node encountered during checking.
+     * @param failureMsg The failure message to be reported.
+     * @param warningMsg The warning message to be reported.
+     */
+    private void checkEnclosingVariable(Tree node, Class<? extends Annotation> targetEffect, String failureMsg, String warningMsg) {
+        VariableTree varTree = TreeUtils.enclosingVariable(getCurrentPath());
+        VariableElement varElt = TreeUtils.elementFromDeclaration(varTree);
+        Class<? extends Annotation> varCallerEffect = atypeFactory.getDeclaredEffect(varElt);
+        if(warningMsg != null) {
+            if(checker.getOption("ignoreWarnings") == null)
```

Comment:
If this check method is going to be called frequently in a codebase with lots of errors, the results of these option lookups should be checked and saved as boolean fields in the constructor, and those boolean fields should be accessed here.  There's no reason to repeat the lookups, since the values won't change while the checker is running.